ipa-core/src/protocol/ipa_prf/malicious_security/verifier.rs

```diff
@@ -95,32 +97,38 @@ pub fn compute_final_sum_share<F: PrimeField, const λ: usize, const P: usize>(z
     (1..λ).fold(F::ZERO, |acc, i| acc + zkp[i])
 }
 
-/// This function compresses the `u_or_v` values and returns the next `u_or_v` values.
+/// This function compresses the input by evaluating the Lagrange table using $lambda$ input values
+/// and packing the results at the beginning of the input vector. For example:
+/// [0, 1, 2, 3, 4, 5, 6, 7, 8, 9] with $lambda$ = 4 and P(0..3) = 10, P(4, 7) = 11, P(8, 9) = 12
+/// will result in input to be set to [10, 11, 12].
 ///
-/// The function uses streams since stream offers a chunk method.
-fn recurse_u_or_v<'a, F: PrimeField, J, const λ: usize>(
-    u_or_v: J,
-    lagrange_table: &'a LagrangeTable<F, λ, 1>,
-) -> impl Iterator<Item = F> + 'a
-where
-    J: Iterator<Item = F> + 'a,
-{
-    u_or_v
-        .chunk_array::<λ>()
-        .map(|x| lagrange_table.eval(&x)[0])
+/// The new length of the input is returned, so the input size can be adjusted accordingly.
+fn compress<const λ: usize, F: PrimeField>(
+    input: &mut [F],
+    lagrange_table: &LagrangeTable<F, λ, 1>,
+) -> usize {
+    let mut write = 0;
+    for start in (0..input.len()).step_by(λ) {
+        let end = min(start + λ, input.len());
+        let chunk = &input[start..end];
+
+        // pad with zeroes if the length of the last chunk is less
+        // than λ.
+        let mut table_input = [F::ZERO; λ];
+        table_input[..chunk.len()].copy_from_slice(chunk);
+
+        input[write] = lagrange_table.eval(table_input)[0];
+        write += 1;
+    }
+
+    write
 }
 
-/// This function recursively compresses the `u_or_v` values.
-/// The recursion factor (or compression) of the first recursion is `λ_FIRST`
-/// The recursion factor of all following recursions is `λ`.
-pub fn recursively_compute_final_check<F: PrimeField, J, const λ_FIRST: usize, const λ: usize>(
-    u_or_v: J,
+pub fn recursively_compute_final_check<F: PrimeField, const λ_FIRST: usize, const λ: usize>(
+    mut u_or_v: Vec<F>,
     challenges: &[F],
     p_or_q_0: F,
-) -> F
-where
-    J: Iterator<Item = F>,
-{
+) -> F {
     let recursions_after_first = challenges.len() - 1;
 
     // compute Lagrange tables
```

Comment:
this probably deserve some changes in separate PR. We don't seem to save anything by collecting all these tables in advance